apmproxy/apmserver.go

```diff
@@ -40,28 +40,40 @@ type jsonError struct {
 	Document string `json:"document,omitempty"`
 }
 
-// ForwardApmData receives agent data as it comes in and posts it to the APM server.
-// Stop checking for, and sending agent data when the function invocation
+// ForwardApmData receives apm data as it comes in and posts it to the APM server.
+// Stop checking for, and sending apm data when the function invocation
 // has completed, signaled via a channel.
-func (c *Client) ForwardApmData(ctx context.Context, metadataContainer *MetadataContainer) error {
+func (c *Client) ForwardApmData(ctx context.Context) error {
 	if c.IsUnhealthy() {
 		return nil
 	}
+
+	// Wait for metadata to be available, metadata will be available as soon as
+	// the first agent data is processed.
+	select {
+	case <-ctx.Done():
+		c.logger.Debug("Invocation context cancelled, not processing any more agent data")
+		return nil
+	case data := <-c.AgentDataChannel:
+		if err := c.forwardAgentData(ctx, data); err != nil {
+			return err
+		}
+	case <-c.metadataAvailable:
+	}
+
+	// Process lambda and agent data after metadata is available
 	for {
 		select {
 		case <-ctx.Done():
 			c.logger.Debug("Invocation context cancelled, not processing any more agent data")
 			return nil
-		case agentData := <-c.DataChannel:
-			if metadataContainer.Metadata == nil {
-				metadata, err := ProcessMetadata(agentData)
-				if err != nil {
-					return fmt.Errorf("failed to extract metadata from agent payload %w", err)
-				}
-				metadataContainer.Metadata = metadata
+		case data := <-c.AgentDataChannel:
+			if err := c.forwardAgentData(ctx, data); err != nil {
+				return err
 			}
-			if err := c.PostToApmServer(ctx, agentData); err != nil {
-				return fmt.Errorf("error sending to APM server, skipping: %v", err)
+		case data := <-c.LambdaDataChannel:
+			if err := c.forwardLambdaData(ctx, data); err != nil {
+				return err
 			}
 		}
 	}
```

Comment:
Do we still need the `metadataAvailable` channel? And if we do, does it need to be threaded through the entire application? IIANM we only care about it in the apmproxy package now.

IIANM, I think we could get rid of the channel and simplify this code at the same time by using a pattern like this:

```go
if c.IsUnhealthy() {
    return nil
}

// Process lambda and agent data after metadata is available
var lambdaDataChannel chan APMData
for {
    select {
    case <-ctx.Done():
        c.logger.Debug("Invocation context cancelled, not processing any more agent data")
        return nil
    case data := <-c.AgentDataChannel:
        if err := c.forwardAgentData(ctx, data); err != nil {
            return err
        }
        // Metadata is now available, so we can start processing events from the Lambda platform.
        lambdaDataChannel = c.LambdaDataChannel
    case data := <-c.LambdaDataChannel:
        if err := c.forwardLambdaData(ctx, data); err != nil {
            return err
        }
    }
}
```

Reply:
Love this solution